posts/nyc-subway-data-with-kafka-streams-and-jhipster-part-1.md

```diff
@@ -0,0 +1,329 @@
+// The interface MtaStream defines the other two bindings as regular channels
+@EnableBinding(value = {KafkaStreamsProcessor.class, MessagingConfiguration.MtaStream.class})
+public class MessagingConfiguration {
+
+    public interface MtaStream {
+        String OUTPUT = "mta-output";
+        String INPUT = "stream-input";
+
+        @Output(OUTPUT)
+        MessageChannel output();
+
+        @Input(INPUT)
+        SubscribableChannel input();
+    }
+
+}
+```
+
+One last thing is to change the file `kafka.yml` to have the two topics created. The environment variable to change is `KAFKA_CREATE_TOPICS`:
+```yml
+KAFKA_CREATE_TOPICS: "mta:1:1,mta-stream:1:1"
+```
+
+# Data processing using Kafka Streams
+
+## Publishing MTA's data
+
+The interface `MtaStream` is used to publish messages to the topic `mta`, it must be injected as a Bean. That's how the method `publishMtaFeeds` will look now:
+```java
+@Scheduled(cron = "0 */10 * * * *")
+public void publishMtaFeeds() {
+	// Feed ids, more details at https://datamine.mta.info/list-of-feeds
+	Stream<String> feedIds = Stream.of("1", "2", "11", "16", "21", "26", "31", "36", "51");
+
+	// Read each feed and build a list of active subways
+	List<Subway> subways = feedIds
+	    .flatMap(this::readMtaFeed)
+	    .collect(Collectors.toList());
+
+	// Publish all subways
+	if (!subways.isEmpty()) {
+	    mtaStream.output().send(MessageBuilder.withPayload(subways).build());
+	}
+}
+```
+
+The message is automatically serialized to a JSON array that contains all active subways.
+
+## Counting subways per line
+
+```java
+@StreamListener("input")
+@SendTo("output")
+public KStream<?, SubwayCount> streamMtaFeeds(KStream<Object, List<Map<String, String>>> input) {
+    // Count subways for each route with a window of 10 mins
+    // Then publish the stream
+    return input
+        .flatMapValues(value -> value.stream().map(subway -> subway.get("route")).collect(Collectors.toList()))
+        .map((key, value) -> new KeyValue<>(value, value))
+        .groupByKey()
+        .windowedBy(TimeWindows.of(10 * 60 * 1000))
+        .count(Materialized.as("subwaycounts"))
+        .toStream()
+        .map((key, value) -> new KeyValue<>(null, new SubwayCount(key, value)));
+}
+```
+
+The above method does all the magic, it streams the topic `mta` and group by the route (which is the line). A window of 10mins is used since the API is polled every 10mins and the stream result is published to the topic `mta-stream`.
+
+The Kafka Streams binder allows to directly use the `KStream` object, the stream processing method will be more clear and simple. The first step is to map the route from the `Subway` class and then map it to a `KeyValue` so it can be grouped by the key. Then the second step defines the window and count the number of records in the stream. And finally the last step converts the stream back to a `KStream` and map a `KeyValue` with a value representing the count:
```

Comment:
```suggestion
The Kafka Streams binder allows to directly use the `KStream` object, the stream processing method will be more clear and simple. The first step is to map the route from the `Subway` class and then map it to a `KeyValue` so it can be grouped by the key. Then the second step defines the window and counts the number of records in the stream. And finally the last step converts the stream back to a `KStream` and map a `KeyValue` with a value representing the count:
```